lib/pkcs11h-core.c

```diff
@@ -703,26 +774,288 @@ pkcs11h_addProvider (
 		sizeof (provider->reference)-1
 	);
 	provider->reference[sizeof (provider->reference)-1] = '\x0';
-	strncpy (
-		provider->manufacturerID,
-		(
-			strlen (provider_location) < sizeof (provider->manufacturerID) ?
-			provider_location :
-			provider_location+strlen (provider_location)-sizeof (provider->manufacturerID)+1
-		),
-		sizeof (provider->manufacturerID)-1
+	provider->manufacturerID[0] = '\x0';
+	provider->provider_location = NULL;
+	provider->allow_protected_auth = FALSE;
+	provider->mask_private_mode = PKCS11H_PRIVATEMODE_MASK_AUTO;
+	provider->slot_event_method = PKCS11H_SLOTEVENT_METHOD_AUTO;
+	provider->slot_poll_interval = 0;
+	provider->cert_is_private = FALSE;
+	provider->init_args = NULL;
+
+	_PKCS11H_DEBUG (
+		PKCS11H_LOG_DEBUG2,
+		"PKCS#11: pkcs11h_registerProvider Provider '%s' manufacturerID '%s'",
+		reference,
+		provider->manufacturerID
+	);
+
+
+#if defined(ENABLE_PKCS11H_THREADING)
+	if ((rv = _pkcs11h_threading_mutexLock (&_g_pkcs11h_data->mutexes.global)) != CKR_OK) {
+		goto cleanup;
+	}
+#endif
+
+	if (_g_pkcs11h_data->providers == NULL) {
+		_g_pkcs11h_data->providers = provider;
+	}
+	else {
+		_pkcs11h_provider_t last = NULL;
+
+		for (
+			last = _g_pkcs11h_data->providers;
+			last->next != NULL;
+			last = last->next
+		);
+		last->next = provider;
+	}
+
+#if defined(ENABLE_PKCS11H_THREADING)
+	_pkcs11h_threading_mutexRelease (&_g_pkcs11h_data->mutexes.global);
+#endif
+
+	rv = CKR_OK;
+
+cleanup:
+
+	_PKCS11H_DEBUG (
+		PKCS11H_LOG_DEBUG1,
+		"PKCS#11: Provider '%s' registered rv=%lu-'%s'",
+		reference,
+		rv,
+		pkcs11h_getMessage (rv)
 	);
-	provider->manufacturerID[sizeof (provider->manufacturerID)-1] = '\x0';
-	provider->allow_protected_auth = allow_protected_auth;
-	provider->mask_private_mode = mask_private_mode;
-	provider->slot_event_method = slot_event_method;
-	provider->slot_poll_interval = slot_poll_interval;
-	provider->cert_is_private = cert_is_private;
 
+	_PKCS11H_DEBUG (
+		PKCS11H_LOG_DEBUG2,
+		"PKCS#11: pkcs11h_registerProvider return rv=%lu-'%s'",
+		rv,
+		pkcs11h_getMessage (rv)
+	);
+
+	return rv;
+}
+
+
+CK_RV
+pkcs11h_setProviderProperty (
+	IN const char * const reference,
+	IN const pkcs11h_provider_property_type_t property,
+	IN const pkcs11h_provider_property_t value
+) {
+#if defined(ENABLE_PKCS11H_THREADING)
+	PKCS11H_BOOL has_mutex_global = FALSE;
+	CK_RV lock_rv;
+#endif
+
+	_pkcs11h_provider_t provider = NULL;
+	CK_RV rv = CKR_OK;
+
+	_PKCS11H_ASSERT (_g_pkcs11h_data!=NULL);
+	_PKCS11H_ASSERT (_g_pkcs11h_data->initialized);
+
+	{
+	const char* property_str = __pkcs11h_property_type_to_str(property);
+	char* value_str = NULL;
+
+
+	if ((rv = __pkcs11h_property_value_to_str(property, value, &value_str)) != CKR_OK)
+	{
+		goto cleanup;
+	}
+
+	_PKCS11H_DEBUG (
+		PKCS11H_LOG_DEBUG2,
+		"PKCS#11: pkcs11h_setProviderProperty entry version='%s', reference='%s', property='%s' value='%s'",
+		PACKAGE_VERSION,
+		reference,
+		property_str,
+		value_str
```

Comment:
all this code is only for debug? maybe just print explicit debug below at switch?

it is risky and inefficient to perform the conversion.